### PR TITLE
Optimize startup time

### PR DESCRIPTION
**Pull request by ghost**
Related to conda-forge/conda-smithy#647.

### CODE REVIEW

**Comment by jakirkham** [comment]
So we want to make sure the `conda-forge` channel is set as higher priority before doing the install. Hence why this was being done in the script.

Ultimately we are just going to want to use our own dev installer instead of Miniconda for this sort of thing. That should give us the largest speedup as we can just start out with packages from `conda-forge` and skip the update step.

I did some work on this not too long ago and called it [miniforge]( https://github.com/jakirkham/miniforge ). This should help all platforms be a bit faster and more consistent. So seems like the better way to go to optimize CIs. We just need someone to explore doing this install in CIs probably at `staged-recipes` to start.

**Comment by ghost** [comment]
The conda-forge channel is already promoted in conda-smithy.

**Comment by ghost** [comment]
If you look at this project's appveyor.yml, you can see that it's promoted here:

https://github.com/conda-forge/conda-forge-build-setup-feedstock/blob/3ad7ac9a83dad6818f9930e0689d9a65515c7b30/appveyor.yml#L44

**Comment by jakirkham** [comment]
At a bare minimum, this shouldn't be OS specific.

**Comment by ghost** [comment]
In principle, I agree with that. However, I think it may be better to roll out this change incrementally. What do you think?

**Comment by jakirkham** [comment]
Sorry I don't follow. What problem do you envision coming up on different OSes?

**Comment by ghost** [comment]
You're correct. There is no reason not to go ahead with this on all platforms.

**Comment by jakirkham** [comment]
As far as `conda-forge` being promoted, this assumes that one is building their feedstocks in `conda-forge`. Historically that has not always been the case, but I have not kept up with other projects doing this.

Edit: Here's an [example org]( https://github.com/GorgonCryoEM ).

**Comment by ghost** [comment]
> As far as conda-forge being promoted, this assumes that one is building their feedstocks in conda-forge. Historically that has not always been the case, but I have not kept up with other projects doing this.

There's no reason that others can't do this in their own appveyor.yml (at least that I can see).

**Comment by ghost** [comment]
We already had the conda.exe bug where an update to the CI files was required.

**Comment by jakirkham** [comment]
In that case, external users wanted to use `conda-smithy` to re-render their own feedstocks. I doubt they would want to hack their CI files afterwards also (assuming I'm understanding your suggestion correctly).

**Comment by ghost** [comment]
What are you talking about? With the linked PR, conda-smithy will regenerate the CI files with the appropriate modifications.

**Comment by ghost** [comment]
P.S. even your linked example doesn't use conda-forge-build-setup from this repository.

**Review by isuruf — APPROVED** [review]

**Comment by ghost** [comment]
@isuruf This should be ready to merge.

**Comment by ghost** [comment]
Ping @jakirkham

**Comment by ghost** [comment]
@isuruf Is this going to be merged? If not, I'll close it in a few days.

File: .travis.yml [hunk 1]
```diff
@@ -49,6 +49,7 @@ install:
       conda config --set show_channel_urls true
       conda install --yes --quiet conda-forge-build-setup
       # Overriding global conda-forge-build-setup with local copy.
+      conda install --yes conda-build-all
```
Comment (isuruf) [inline .travis.yml:52]: Why is this needed?

**Comment by isuruf** [comment]
@conda-forge/conda-forge-build-setup, can I get another review here?

File: .travis.yml (hunk 1 above)
Comment (ghost) [reply .travis.yml:52]: This is needed to bootstrap this package. The `conda-forge-build-setup` installed here doesn't include the additional requirements.

File: recipe/meta.yaml [hunk 2]
```diff
@@ -15,6 +15,16 @@ build:
...
 
+requirements:
+  run:
+    - conda
+    - conda-env
+    - conda-build
+    - jinja2
+    - anaconda-client
+    - conda-build ==2
+    - vs2008_express_vc_python_patch  # [win]
```
Comment (patricksnape) [inline recipe/meta.yaml]: `and py27`?

Comment (isuruf) [reply recipe/meta.yaml]: We don't know which python version is going to be used in the recipe, so this is installed everywhere for windows as is the case in master

**Comment by ghost** [comment]
@isuruf Is this moving forward?

**Comment by patricksnape** [comment]
Looks fine to me but probably wants @jakirkham or someone who worked directly on this package before to comment

File: .travis.yml (hunk 1 above)
Comment (jakirkham) [reply .travis.yml:52]: Why is this needed to bootstrap the package? Not following. 😕

File: recipe/meta.yaml [hunk 3]
```diff
@@ -15,6 +15,16 @@ build:
...
     - cp "${RECIPE_DIR}/upload_or_check_non_existence.py" "${PREFIX}/bin/upload_or_check_non_existence"                  # [unix]
 
+requirements:
+  run:
+    - conda
+    - conda-env
+    - conda-build
+    - jinja2
+    - anaconda-client
+    - conda-build ==2
```
Comment (jakirkham) [inline recipe/meta.yaml]: We probably shouldn't be listing this twice (i.e. it shows up above too).

File: recipe/meta.yaml (hunk 2 above)
Comment (jakirkham) [reply recipe/meta.yaml]: Yeah, this is actually no different than what we are already doing here. Though agree it would be nice if we didn't always have to do it. Think it has been easier for everyone to just leave it as is.

**Comment by jakirkham** [comment]
Sorry have been busy with things IRL.

Made a few comments above.

Generally don't mind these changes as long as `conda-smithy` gets similar changes.

Would note that we are preparing for a `conda-smithy` 3.0.0 release that relies on `conda-build` 3. So we will need to finish that before we can make a `conda-smithy` 3.0.0 release that works with these changes. Alternatively if it is important to do soon, we can backport the `conda-smithy` PR (once it is accepted) and do a patch release 2.4.6.

**Comment by isuruf** [comment]
> Generally don't mind these changes as long as conda-smithy gets similar changes.

What kind of changes are needed in conda-smithy?

**Comment by jakirkham** [comment]
If `conda-build-all` is actually needed to be installed in CI, we need this change in `conda-smithy`. Though I'm not convinced that it is needed.

**Comment by ghost** [comment]
`conda-build-all` doesn't need to be installed on the CI. It's only needed here because we need to bootstrap. To build this package, we're installing the previously-built version, which doesn't have the necessary requirements.

**Comment by ghost** [comment]
The `conda-smithy` changes are a minor optimization on top of this, but are not required.

**Comment by isuruf** [comment]
> To build this package, we're installing the previously-built version, which doesn't have the necessary requirements.

Got it. So the upload script will not have `conda-build, anaconda-client` installed in this feedstock only, but in other feedstocks it will work because this package pulls them as well.

File: recipe/meta.yaml [hunk 4]
```diff
@@ -1,6 +1,6 @@
 package:
   name: conda-forge-build-setup
-  version: 4.4.17
+  version: 4.4.18
```
Comment (jakirkham) [inline recipe/meta.yaml:3]: In retrospect, it might have been better to do `5.0.0` given the bootstrapping issue. Though it's not that big of a deal. We will likely go to `5.0.0` soon enough with `conda-build`.

**Comment by isuruf** [comment]
Moved the packages built here to a broken label. See https://github.com/conda-forge/libgit2-feedstock/pull/15

**Comment by isuruf** [comment]
@kalefranz, said
> That looks like conda/conda#6057, which was fixed in conda/conda#6277, released in 4.3.32.

We can do this when conda-smithy 3.0.0 is released.
